// @rhuds/core/src/theme/index.ts
export { ThemeProvider } from './ThemeProvider'
export { useTheme } from './useTheme'
export { createTheme } from './createTheme'
export { ThemeContext } from './context'
export { darkMode, lightMode, neonGreenMode, neonBlueMode, neonRedMode, DEFAULT_THEMES } from './themes'
export type { ThemeTokens, ThemeMode, ThemeContextValue } from './types'

// Export RHUDS Pro theme creation functions
export {
  createThemeUnit,
  createThemeColor,
  createThemeStyle,
  createThemeBreakpoints,
  createCreateTheme,
  createAppTheme,
} from './creators'

// Export RHUDS Pro color manipulation utilities (Requirements 2.1-2.4)
export {
  // Color conversion
  hexToRgb,
  rgbToHex,
  rgbToHsl,
  hslToRgb,
  
  // Color manipulation (Requirement 2.1)
  lighten,
  darken,
  saturate,
  desaturate,
  generateColorVariations,
  
  // Alpha channel (Requirement 2.2)
  alpha,
  createAlphaFunction,
  
  // Gradients (Requirement 2.3)
  createGradient,
  createLinearGradient,
  createRadialGradient,
  createConicGradient,
  
  // Animated transitions (Requirement 2.4)
  interpolateColor,
  createColorTransitionKeyframes,
  createColorTransitionAnimation,
  animateGradient,
  
  // Utilities
  isValidHexColor,
  parseColor,
} from './colorUtils'

export type {
  ColorTransitionKeyframe,
  ColorTransitionConfig,
} from './colorUtils'

// Export RHUDS Pro theme validation
export {
  validateTheme,
  isValidTheme,
  ThemeValidationError,
} from './validation'

// Export RHUDS Pro theme models
export type {
  // Main theme interface
  RHUDSTheme,
  ThemeConfig,
  ThemeOverride,
  DeepPartial,
  
  // Color system
  ColorPalette,
  ColorSystem,
  GradientDefinition,
  RGB,
  HSL,
  VariationOptions,
  
  // Unit system
  UnitScale,
  UnitSystem,
  ShadowScale,
  
  // Typography system
  TypographySystem,
  FontFamilySystem,
  FontSizeScale,
  FontWeightScale,
  LineHeightScale,
  LetterSpacingScale,
  
  // Breakpoint system
  BreakpointSystem,
  BreakpointValues,
  BreakpointLabels,
  
  // Animation system
  AnimationDefaults,
  DurationScale,
  EasingFunctions,
  
  // Z-index system
  ZIndexSystem,
  
  // Serialization
  SerializableTheme,
} from './models'

// Export theme management (Requirements 1.8-1.10, 51.1-51.7)
export {
  ThemeManager,
  LocalStorageThemeStorage,
  ThemeSyncManager,
  extendTheme,
  composeThemes,
  getSystemThemePreference,
  watchSystemThemePreference,
} from './ThemeManager'

export type { ThemeStorage } from './ThemeManager'

export { useThemeManager } from './useThemeManager'
