// @rhuds/core/src/store/index.ts
export { store } from './store'
export type { RootState, AppDispatch } from './store'

// Theme slice
export { default as themeReducer } from './slices/themeSlice'
export {
  setTheme,
  customizeToken,
  resetTheme,
  loadThemeFromStorage,
} from './slices/themeSlice'
export type { ThemeState } from './slices/themeSlice'

// UI slice
export { default as uiReducer } from './slices/uiSlice'
export {
  openModal,
  closeModal,
  closeAllModals,
  openDropdown,
  closeDropdown,
  closeAllDropdowns,
  addNotification,
  removeNotification,
  clearNotifications,
} from './slices/uiSlice'
export type { UIState, Notification } from './slices/uiSlice'

// SFX slice
export { default as sfxReducer } from './slices/sfxSlice'
export {
  setEnabled,
  setVolume,
  toggleEnabled,
  loadSFXFromStorage,
} from './slices/sfxSlice'
export type { SFXState } from './slices/sfxSlice'

// Animation slice
export { default as animationReducer } from './slices/animationSlice'
export {
  setAnimationEnabled,
  setReducedMotion,
  setGlobalSpeed,
  registerAnimation,
  unregisterAnimation,
  setAnimationDuration,
  setAnimationEasing,
  loadAnimationFromStorage,
} from './slices/animationSlice'
export type { AnimationState } from './slices/animationSlice'

// Audio slice
export { default as audioReducer } from './slices/audioSlice'
export {
  setMasterVolume,
  setMuted,
  toggleMute,
  setCategoryVolume,
  setCategoryMuted,
  toggleCategoryMute,
  registerBleep,
  unregisterBleep,
  setSpatialAudioEnabled,
  loadAudioFromStorage,
} from './slices/audioSlice'
export type { AudioState } from './slices/audioSlice'

// Middleware
export {
  createPersistenceMiddleware,
  loadPersistedState,
} from './middleware/persistence'
export type { PersistenceConfig } from './middleware/persistence'

export {
  createUndoRedoMiddleware,
  undo,
  redo,
  clearHistory,
  canUndo,
  canRedo,
} from './middleware/undoRedo'
export type { UndoRedoConfig } from './middleware/undoRedo'
