// @rhuds/core/src/store/store.ts
import { configureStore } from '@reduxjs/toolkit'
import themeReducer from './slices/themeSlice'
import uiReducer from './slices/uiSlice'
import sfxReducer from './slices/sfxSlice'
import animationReducer from './slices/animationSlice'
import audioReducer from './slices/audioSlice'
import { createPersistenceMiddleware, loadPersistedState } from './middleware/persistence'
import { createUndoRedoMiddleware } from './middleware/undoRedo'

// Load persisted state
const persistedState = loadPersistedState()

export const store = configureStore({
  reducer: {
    theme: themeReducer,
    ui: uiReducer,
    sfx: sfxReducer,
    animation: animationReducer,
    audio: audioReducer,
  },
  preloadedState: persistedState,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        // Ignore notification action callbacks
        ignoredActions: ['ui/addNotification'],
        ignoredPaths: ['ui.notifications[].action.onClick'],
      },
    })
      .concat(
        createPersistenceMiddleware({
          whitelist: ['theme', 'animation', 'audio'],
          throttle: 1000,
        })
      )
      .concat(
        createUndoRedoMiddleware({
          limit: 50,
          ignoreActions: ['ui/addNotification', 'ui/removeNotification'],
        })
      ),
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
