// @rhuds/hooks/src/index.ts
// Theme hooks
export { useTheme } from './useTheme'

// Animation hooks
export { useComponentAnimation } from './useComponentAnimation'
export { useAnimation } from './useAnimation'
export { useAnimator } from './useAnimator'
export { useSpring } from './useSpring'

// Audio hooks
export { useBleeps } from './useBleeps'
export { useAudioVisualization } from './useAudioVisualization'

// Form hooks
export { useFormField } from './useFormField'

// Utility hooks
export { useLocalStorage } from './useLocalStorage'
export { useSessionStorage } from './useSessionStorage'
export { useDebounce } from './useDebounce'
export { useThrottle } from './useThrottle'
export { usePrevious } from './usePrevious'
export { useAsync } from './useAsync'
export { useMediaQuery } from './useMediaQuery'
export { useClickOutside } from './useClickOutside'
export { useKeyPress } from './useKeyPress'

// Types
export type { UseComponentAnimationOptions, UseComponentAnimationReturn, AnimationTrigger } from './useComponentAnimation'
export type { UseAnimationOptions, UseAnimationReturn } from './useAnimation'
export type { AnimatorState, UseAnimatorReturn } from './useAnimator'
export type { UseSpringOptions, UseSpringReturn } from './useSpring'
export type { PlayOptions, UseBleepsReturn } from './useBleeps'
export type { UseAudioVisualizationReturn } from './useAudioVisualization'
export type { FormValidationRule, UseFormFieldReturn } from './useFormField'
export type { UseAsyncReturn } from './useAsync'
